### PR TITLE
Oculo code review

### PR DESCRIPTION
**Pull request by anhnk**
# Oculo Code Review
## The Code
Inside of `code.zip` is a simple controller from a bigger application that allows users to send communications to each other.
The functionallity can be described as:
- Allow an API client to create a message from an existing user to an existing user and send it via an external api
- Allow an API client to query the message at a later date to determine if the message has been successfully sent.

The code also contains a `ExternalApiClient` that represents an interface for making calls to an external system which can
take varying amounts of time to succeed, fail and raise errors (like timeouts).

It also contains an integration spec that tests the desired behaviour

It's also worth noting this code has been designed to be problematic and is not indicative of how we write code at Oculo

## The Exercise
We would like you to do the following:
- Create a repository called `oculo_code_review`
- Download and extract the contents of `code.zip`
- Create a pull request against your repo with the contents of `code.zip`
- Review the pull request identifying problems you find in the code and suggest changes you would need to make to get the code
  to a **production** ready standard.
- When examining the code you can assume that the rest of the application is coded in a similar way so will also be problematic.
- Provide commentary on any considerations or questions that might need to be answered about the code.
- You can safely ignore any issues that a linter could auto-correct.

When you're happy with your review send us a link to the PR.

### CODE REVIEW

File: server.rb [hunk 1]
```diff
@@ -0,0 +1,189 @@
+require "sinatra/base"
+require "sqlite3"
+require "nokogiri"
+require 'active_support'
+require 'json'
+require 'securerandom'
```
Comment (anhnk) [inline server.rb:6]: It it better to be consistent with the usage of single quotes `'` and double quotes `"`. Most Ruby linters suggest using single quotes for strings and double ones for interpolation/escaping sequences.

Besides, it looks like `active_support` is not being used here so it can be removed. The gem should be removed from `Gemfile` and `Gemfile.lock` as well.

File: Gemfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+source 'https://rubygems.org'
+
+gem 'sinatra'
+gem 'sqlite3'
+gem 'nokogiri'
+gem 'activesupport'
```
Comment (anhnk) [inline Gemfile:8]: `activesupport` gem doesn't seem to be used in this project.

File: spec_helper.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+require "rack/test"
+require "securerandom"
+require 'securerandom'
+require 'nokogiri'
```
Comment (anhnk) [inline spec_helper.rb:4]: Same comment regarding the usage and single and double quotes.

Besides, `securerandom` and `nokogiri` should NOT be required here. They should be in the code file, which is `server.rb` in this case.

File: server.rb [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+class Server < Sinatra::Base
```
Comment (anhnk) [inline server.rb:8]: This file is quite long and hard to follow because it also has the implementation of classes such as `JSONMessage`, `XMLMessage`, `ExternalApiClient`. Each of these classes should be put into their own separate file to keep the code modular and concise. This will make writing unit tests a lot easier since modules are isolated and stubs/mocks can be used.

Please see some of the comments below regarding tests.

File: server_spec.rb [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+      end
+    end
+
+    context 'other' do
+      let(:format) { 'csv' }
+      specify { expect(response.status).to eq(400) }
+    end
+
+  end
+end
```
Comment (anhnk) [inline server_spec.rb:234]: This integration spec looks good. However, we are missing unit tests here. 

Tests can be grouped into buckets of different granularity. In contrast to integration tests, unit tests can run faster and allow more isolation regarding the component we would like to tests. In this case, we can test more edge cases with `JSONMessage`, `XMLMessage`, and `ExternalApiClient` classes.

File: server.rb [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+          end
+        end
+        [200, builder.to_xml]
+      else
+        [400]
+      end
+    else
+      [400]
+    end
+  end
```
Comment (anhnk) [inline server.rb:47]: When our API returns a client error, it would be more useful to provide a bit more information about the error, whether the format is not acceptable or the params/request body is invalid.

We should also return an appropriate server error with `500` error code when the `ExternalApiClient` fails to send the message.

This block of code seems to have high level of complexity with many levels of indentation. Personally, I'd like to write these `controller`-ish blocks of code with around 10 lines and move most of the logic into the service layer. It's probably more of a personal preference but I do find it a lot more readable and the controller files can be kept very short.

An example is:
```ruby
post '/messages/:format?' do
  format = params['format']
  return [400, {error: 'The acceptable formats are json, xml.'}] unless ['json', 'xml'].include?(format)

  message = Message.parse(request.body.read, format)
  return [400, ResponseBuilder.invalid_request(format)] unless Message.valid?(message)
  
  id = Message.save(message)
  sent = Message.send(message)
  return [500, ResponseBuilder.failed_message(format)] unless sent

  [200, ResponseBuilder.sent_message(format, id)]
end
```

File: server.rb [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+          xml.message do
+            xml.status status.flatten.first
+          end
+        end
+      end
+      [200, builder.to_xml]
+    else
+      [400]
+    end
+  end
```
Comment (anhnk) [inline server.rb:67]: Similar to the comment above, this can be improved as follows:

```ruby
get '/messages/:id/status/:format?' do
  format = params['format']
  return [400, {error: 'The acceptable formats are json, xml.'}] unless ['json', 'xml'].include?(format)

  status = db.execute("SELECT status from messages WHERE id = '#{params['id']}' LIMIT 1")
  return [404, ResponseBuilder.not_found(format) if status.empty?

  [200, ResponseBuilder.message_status(format, status)]
end
```

File: server.rb [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+  get '/messages/:id/status/:format?' do
+    status = db.execute("SELECT status from messages WHERE id = '#{params['id']}' LIMIT 1")
+    return [400] if status.empty?
+    case params['format']
+    when 'json'
+      [200, {message: { status: status.flatten.first } }.to_json ]
+    when 'xml'
+      builder = Nokogiri::XML::Builder.new do |xml|
```
Comment (anhnk) [inline server.rb:56]: Nokogiri is a huge native depedency. [Oga](https://github.com/yorickpeterse/oga) is a much more lightweight alternative for an XML/HTML parser with similar interface. It is probably worth a try. :)

File: server.rb [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        end
+      end
+      [200, builder.to_xml]
+    else
+      [400]
+    end
+  end
+
+  def db
+    SQLite3::Database.new "test.db"
```
Comment (anhnk) [inline server.rb:70]: `SQlite3` is probably easy to set up and get started but for the long term project we may one day want to consider using something with better features such as Postgres or MySQL.

File: server.rb [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+          WHERE id = '#{@id}';
+        SQL
+                  )
+      end
+    end
+
+    def db
+      SQLite3::Database.new "test.db"
+    end
+  end
```
Comment (anhnk) [inline server.rb:178]: `JSONMessage` and `XMLMessage` classes have a lot of duplicate code. The only difference is how the input is parsed. I would suggest having `Message` class responsible for validating, saving, and sending the message and using a separate class like `MessageParser` to parse the message based on the request/response format. See the comment regarding how the controllers are written for this to make more sense.

File: server.rb [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+      parsed_json = JSON.parse(data)
+      [parsed_json['to'], parsed_json['from'], parsed_json['message']]
+    end
+
+    def valid?
+      return false if [@to, @from, @message].any?(&:nil?)
+      return false if db.execute("SELECT * from users where id = '#{@to}' LIMIT 1").empty?
+      return false if db.execute("SELECT * from users where id = '#{@from}' LIMIT 1").empty?
+      true
+    end
```
Comment (anhnk) [inline server.rb:89]: Nice. I really like how Ruby idioms are being used here!

File: server.rb [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+
+class Server < Sinatra::Base
+
+  # {
+  #  to: <UUID>
+  #  from: <UUID>
+  #  message: <String>
+  # }
+
+  post '/messages/:format?' do
```
Comment (anhnk) [inline server.rb:16]: Both of the `post` and `get` actions don't have any authentication/authorisation steps. I guess it is alright to keep things simple but for security we may want to consider adding this in the future.

File: server.rb [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+                  )
+      else
+        db.execute(<<-SQL
+          UPDATE messages
+          SET status = 'failed'
+          WHERE id = '#{@id}';
+        SQL
+                  )
+      end
+    end
```
Comment (anhnk) [inline server.rb:118]: This could be shortened as:

```ruby
def send_message
  sent = ExternalApiClient.send_message({to: @to, from: @from, message: @message})
  status = sent ? 'sent' : 'failed'
  db.execute(<<-SQL
          UPDATE messages
          SET status = '#{@status}'
          WHERE id = '#{@id}';
        SQL
                  )
end
```

The example I provide here is still using the instance variables `@to`, `@from`, `@message`. But if we are going to follow the suggestion above in the controllers, these arguments will need to be passed in the method.

File: server.rb [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+      true
+    end
+
+    def save
+      @id = SecureRandom.uuid
+      db.execute(<<~SQL
+        INSERT INTO messages(id, to_user, from_user, message, status)
+        VALUES('#{@id}', '#{@to}','#{@from}', '#{@message}', 'pending');
+        SQL
+      )
```
Comment (anhnk) [inline server.rb:97]: I am a little concerned about how the values are inserted here. Since I usually use Rails helper most of the time, I can't confirm if this is adequate to prevent injection attacks. Maybe we would want to have better sanitisation to increase our security measure. Let me know if this is safe enough.

**Review by anhnk — COMMENTED** [review]
Please see my comments below.

To summarise, we can break the big file `server.rb` into smaller ones and add more unit tests. A few refactoring points can be done such as creating a class to be responsible for parsing the request input and thus removing duplicate code, adding a response builder to generate the response to the client (json or xml).

Please let me know if you need more explanation.
